[clang][dataflow] Add parameters of other functions to `LambdaCaptured…

bazuzi opened the pull request:
…Params`.

"Other functions" here are functions other than the FunctionDecl provided to `getReferencedDecls`. Adding the parameters to `LambdaCapturedParams` because the only known triggering circumstance is a synthetic Stmt used for dataflow analysis of lambda init-captures where the initializer references the parameter(s) of the function surrounding the LambdaExpr.

bazuzi committed:
[clang][dataflow]Add parameters of other functions to `LambdaCapturedParams`.

"Other functions" here are functions other than the FunctionDecl provided to `getReferencedDecls`. Adding the parameters to `LambdaCapturedParams` because the only known triggering circumstance is a synthetic Stmt used for dataflow analysis of lambda init-captures where the initializer references the parameter(s) of the function surrounding the LambdaExpr.

ymand commented on clang/include/clang/Analysis/FlowSensitive/ASTOps.h:
```diff
@@ -154,6 +154,14 @@ struct ReferencedDecls {
   /// the surrounding function) that the lambda captures. Captured local
   /// variables are already included in `Locals` above.
+  ///
+  /// This set also includes any referenced parameters of functions other than
```
This wording feels general (and hence is a bit hard to parse). Could you be more specific, like

"of the surrounding function that introduces  the lambda whose body is targeted for visitation"

? Or, is there actually more general functionality here?

bazuzi committed:
Address review comments and add another test.

Inline a function instead of needing a good name for it, rewrite
explanatory comments, and add a test to demonstrate lack of behavior
change for common circumstances.

bazuzi commented on clang/include/clang/Analysis/FlowSensitive/ASTOps.h:
```diff
@@ -154,6 +154,14 @@ struct ReferencedDecls {
   /// When analyzing a lambda's call operator, the set of all parameters (from
   /// the surrounding function) that the lambda captures. Captured local
```
It doesn't do those things when analyzing a lambda's call operator. We don't see the init-captures when visiting the call operator.

ymand commented on clang/include/clang/Analysis/FlowSensitive/ASTOps.h:
```diff
@@ -154,6 +154,17 @@ struct ReferencedDecls {
+  /// referenced function parameters. This supports the collection of
+  /// ReferencedDecls from a DeclStmt constructed for lambda init-capture
+  /// VarDecls for the purpose of performing a dataflow analysis on the
+  /// declaration/initialization. When analyzing any `FunctionDecl`, this set
```
super nit: I might preface with "Note:" to highlight that this is an aside and no longer describing the variable.

ymand reviewed (approved):
Thanks!
